### PR TITLE
[BEAM-1988] Add FileSystems Interface for accessing underlying FS

### PR DESCRIPTION
**Pull request by sb2nov**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

R: @chamikaramj @aaltay PTAL

### CODE REVIEW

File: sdks/python/apache_beam/io/filesystems.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+  @staticmethod
+  def create(path, mime_type='application/octet-stream',
+             compression_type=CompressionTypes.AUTO):
+    """Returns a write channel for the given file path.
+
+    Args:
+      path: string path of the file object to be written to the system
+      mime_type: MIME type to specify the type of content in the file object
+      compression_type: Type of compression to be used for this object
```
Comment (sb2nov) [inline sdks/python/apache_beam/io/filesystems.py]: Done.

File: sdks/python/apache_beam/io/filesystems.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+    Returns: file handle with a close function for the user to use
```
Comment (sb2nov) [inline sdks/python/apache_beam/io/filesystems.py]: Done.

File: sdks/python/apache_beam/io/filesystems.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  def open(path, mime_type='application/octet-stream',
+           compression_type=CompressionTypes.AUTO):
+    """Returns a read channel for the given file path.
+
+    Args:
+      path: string path of the file object to be written to the system
+      mime_type: MIME type to specify the type of content in the file object
+      compression_type: Type of compression to be used for this object
+
+    Returns: file handle with a close function for the user to use
```
Comment (sb2nov) [inline sdks/python/apache_beam/io/filesystems.py]: Done.

File: sdks/python/apache_beam/io/filesystems_test.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+  def tearDown(self):
+    shutil.rmtree(self.tmpdir)
+
+  @mock.patch('apache_beam.io.localfilesystem.os')
+  def test_unix_path_join(self, *unused_mocks):
+    # Test joining of Unix paths.
+    localfilesystem.os.path.join.side_effect = _gen_fake_join('/')
+    self.assertEqual('/tmp/path/to/file',
+                     FileSystems.join('/tmp/path', 'to', 'file'))
```
Comment (sb2nov) [inline sdks/python/apache_beam/io/filesystems_test.py]: Done.

File: sdks/python/apache_beam/io/filesystems_test.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    self.assertEqual('/tmp/path/to/file',
+                     FileSystems.join('/tmp/path', 'to/file'))
+
+  @mock.patch('apache_beam.io.localfilesystem.os')
+  def test_windows_path_join(self, *unused_mocks):
+    # Test joining of Windows paths.
+    localfilesystem.os.path.join.side_effect = _gen_fake_join('\\')
+    self.assertEqual(r'C:\tmp\path\to\file',
```
Comment (sb2nov) [inline sdks/python/apache_beam/io/filesystems_test.py]: Done.

File: sdks/python/apache_beam/io/filesystems.py (hunk 1 above)
Comment (sb2nov) [inline sdks/python/apache_beam/io/filesystems.py]: Done.

**Comment by sb2nov** [comment]
R: @chamikaramj PTAL

**Comment by chamikaramj** [comment]
Retest this please

**Comment by sb2nov** [comment]
Guess Jenkins is still not back, Travis seems happy though.

**Comment by chamikaramj** [comment]
Sg. Will merge after some manual testing.

**Comment by sb2nov** [comment]
Thanks

**Comment by chamikaramj** [comment]
LGTM (for got to do that :) )

**Pull request closed without merging**